A39-xds-http-filters.md

```diff
@@ -0,0 +1,165 @@
+  may not be able to handle.
+
+When gRPC receives the `HttpConnectionManager` proto in the LDS
+response, it will validate the list of filters as follows:
+- Every filter in the list must be of a known type, as identified by the
+  `type_url` field.  Any unknown config type will cause the Listener
+  resource to be NACKed.
+- The fields in the filter config should be validated by the filter
+  implementation.  Any validation error will cause the Listener resource
+  to be NACKed.
+- Every filter in the list must have a unique instance name.  Any
+  duplicate instance names will cause the Listener resource to be NACKed.
+
+Note that gRPC will support configuring multiple instances of the same
+filter implementation, each with its own name and config.
+
+In each of the `VirtualHost`, `Route`, and `WeightedCluster.ClusterWeight`
+protos, there is a field called `typed_per_filter_config`, which is a map
+from filter instance name to a corresponding filter-specific configuration
+proto message.  (Note that Envoy currently keys these maps by legacy
+filter names instead of by filter instance names, but this is expected
+to change, as per https://github.com/envoyproxy/envoy/issues/12274; gRPC
+will directly implement the desired semantics here.)
+
+When gRPC receives the `RouteConfiguration` proto, either in the LDS or
+RDS response, it will validate the contents of each
+`typed_per_filter_config` map as follows:
+- Every config in the list must be of a known type, as identified by the
+  map value proto message type.  Any unknown config type will cause the
+  resource to be NACKed.
+- The fields in the filter config should be validated by the filter
+  implementation.  Any validation error will cause the resource to be NACKed.
+- Note that gRPC will *not* fail validation if the map key specifies a
+  filter instance name that does not exist in the `HttpConnectionManager`
+  filter list.  This is because during an update, the xDS client code
+  cannot know which `HttpConnectionManager` config is currently being
+  used.
+
+Unlike Envoy, which provides an API that filters must call to access their
+per-{VirtualHost,Route,ClusterWeight} config overrides, gRPC will
+require filters to construct a merged configuration, applying any
+necessary overrides, when the xDS config is applied.  The filters will
+then use this merged configuration at run-time.  However, this approach
+is an implementation detail that could change in the future.
+
+Note that, for a given request, the filter's configuration will come
+from the top-level filter list in the `HttpConnectionManager` config
+and the most specific override.  In other words, if a given filter
+instance has an override in the `ClusterWeight` proto, that will be
+used; otherwise, if it has an override in the `Route` proto, that will
+be used; otherwise, if it has an override in the `VirtualHost` proto,
+that will be used; otherwise, no override will be used.
```

Comment:
This sounds like a limitation vs. the Envoy API which allows a filter to query for any and all of the three overrides.  Is this okay?  I could imagine a system where the config is:

- LDS (partial config that applies to all virtual hosts behind this listener)
- Virtual Host (more config that applies to all routes behind this virtual host)
- Route (more config that applies to all weighted clusters behind this route)
- WeightedCluster (the remaining config that applies to this cluster)

Reply:
The behavior described here is exactly how the Envoy implementation works.  Filters don't have APIs that let them access each override location independently; they just have an API that says "give me the most specific override config".

Reply:
My mistake, then.  This made it sound like they could access them independently:

> Unlike Envoy, which provides an API that filters must call to access their
per-{VirtualHost,Route,ClusterWeight} config overrides